source/opt/licm_pass.cpp

```diff
@@ -0,0 +1,324 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "licm_pass.h"
+#include "cfg.h"
+#include "module.h"
+
+#include "pass.h"
+
+namespace spvtools {
+namespace opt {
+
+LICMPass::LICMPass(){};
+
+Pass::Status LICMPass::Process(ir::IRContext* context) {
+  bool modified = false;
+
+  if (context != nullptr) {
+    ir_context = context;
+    modified = ProcessIRContext();
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool LICMPass::ProcessIRContext() {
+  bool modified = false;
+  ir::Module* module = ir_context->module();
+
+  // Process each function in the module
+  for (ir::Function& f : *module) {
+    modified |= ProcessFunction(&f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessFunction(ir::Function* f) {
+  bool modified = false;
+  ir::LoopDescriptor loop_descriptor{f};
+
+  // dom_analysis = ir_context->GetDominatorAnalysis(f, cfg);
+
+  // Process each loop in the function
+  for (ir::Loop& loop : loop_descriptor) {
+    modified |= ProcessLoop(loop, f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessLoop(ir::Loop& loop, ir::Function* f) {
+  // Process all nested loops first
+  for (ir::Loop* nested_loop : loop) {
+    ProcessLoop(*nested_loop, f);
+  }
+
+  std::vector<ir::Instruction*> instructions{};
+  GatherAllLoopInstructions(loop, &instructions);
+
+  return ProcessInstructionList(loop, &instructions);
+}
+
+void LICMPass::GatherAllLoopInstructions(
+    ir::Loop& loop, std::vector<ir::Instruction*>* instructions) {
+  for (uint32_t bb_id : loop.GetBlocks()) {
+    ir::BasicBlock* bb = ir_context->get_instr_block(bb_id);
+    for (ir::Instruction& inst : *bb) {
+      if (!DoesOpcodeHaveSideEffects(inst.opcode())) {
+        instructions->push_back(&inst);
+      }
+    }
+  }
+}
+
+void LICMPass::HoistInstruction(ir::BasicBlock* pre_header_bb,
+                                ir::Instruction* inst) {
+  ir::InstructionList inst_list{};
+  inst_list.push_back(std::unique_ptr<ir::Instruction>(inst));
+  auto pre_header_branch_inst_it = pre_header_bb->tail();
+
+  pre_header_branch_inst_it.MoveBefore(&inst_list);
+  ir_context->set_instr_block(inst, pre_header_bb);
+}
+
+bool LICMPass::AllOperandsOutsideLoop(ir::Loop& loop,
+                                         ir::Instruction* inst) {
+  for (ir::Operand& operand : *inst) {
+    if (operand.type == SPV_OPERAND_TYPE_ID) {
+      if (loop.IsInsideLoop(
+              ir_context->get_def_use_mgr()->GetDef(operand.words.front()))) {
+        return false;
+      }
+    }
+  }
+  return true;
+}
+
+bool LICMPass::ProcessInstructionList(
+    ir::Loop& loop, std::vector<ir::Instruction*>* instructions) {
+  std::queue<ir::Instruction*> instruction_queue{};
```

Comment:
Oh I see, I have been wrong all this time ... I think it is better to use the queue right at the beginning